wordpress-mobile/WordPress-Android
# Integrating Zendesk

Closes #7737. This PR replaces Helpshift with Zendesk. All of the changes has been reviewed and tested in the following PRs:
#7739, #7756, #7807, #7825, #7829, #7845, #7904, #7905, #7911, #7912, #7933, #7937, #7951.

**Testing Instructions:**

Here is some preliminary information before you start testing Zendesk:

* You can access Zendesk features through "Me" -> "Help & Support", this will be referred as Help page.
* There are 3 Zendesk specific features implemented: "Ticket Creation (Contact us)", "Help Center (Browse our FAQ)", "Ticket List (My Tickets)" and they are all in the Help page.
* There are 3 different ways to create a new ticket. "Contact us" is the most direct way, but there is a floating action button within both My Tickets and FAQ pages.
* Ticket creation within FAQ is only available if the support identity is created. (more on that later)
* The UI for all these 3 features are implemented in the Zendesk SDK. We don't control the behavior or the style of it.
* The localization of Zendesk pages are handled by Zendesk SDK. You'll need to change your device language for it. Changing the language within the app will have no effect.
* When a push notification is received from Zendesk, it'll either refresh the ticket page if it's currently being displayed or it'll create a new notification. Since they always send the same title and message for the push notifications, we use our own messaging for it. The message will not contain anything specific about the ticket.

**How identity works:**

* When the user tries to access "Contact us" or "My Tickets" for the first time, it'll ask them to enter an email and a name. This is independent of any account and must not be set to a Zendesk agent email. We do our best to pre-fill this information.
* Users can change their support email at any time from the "Contact email" field in the Help page. If there is no identity set, this will simply create a new one. That means if the user changes the "Contact email" before they access the "Contact us" or "Browse our FAQ" for the first time, they'll not be asked for their support email and name. The `name` field doesn't really matter, so we accept not having a name set in such a case.
* Due to the way push notifications work in the back end, logging out will reset the users' support identity.
* **Changing the identity in any way will reset the tickets on the device.** This is because we are using an anonymous identity. We have a project proposal to use [JWT based authentication](https://developer.zendesk.com/embeddables/docs/android-support-sdk/sdk_set_identity#setting-a-unique-identity) which will fix this issue.

**Tests:**

1. Create a ticket from all 3 Zendesk pages. Verify the correctness of following fields in Zendesk: appVersion, blogList, deviceFreeSpace, logs, networkInformation, currentSite. There is currently a bug about the device free space issue, so the value might not be correct. #7777 
2. While on a ticket page on your device, reply to it from Zendesk and verify that the reply shows up automatically. You shouldn't get any push notifications for this case.
3. While not in a ticket's page on your device, reply to it from Zendesk and verify that you get a push notification for it. Tapping on the notification should take you to the "Me" page. We are still looking into adding deep-linking to it.
4. Verify that you can see your tickets in the "My tickets" page.
5. Check out the FAQ page and verify that it's working as expected.
6. Change your contact email and verify that you can't access your previous tickets from the device anymore. You should still receive emails for the previous tickets if there are any replies to it.
7. Do a fresh installation and verify that you can access FAQ without setting a support identity.
